File: Examples/PinchCanvas/PinchCanvas/ViewController.m
//
//  ViewController.m
//  PinchCanvas
//
//  Created by Jun on 11/27/14.
//  Copyright (c) 2014 Jun Tanaka. All rights reserved.
//

#import "ViewController.h"


@interface ViewController ()

@property (weak, nonatomic) IBOutlet UIImageView *imageView;
@property (weak, nonatomic) IBOutlet UILabel *textLabel;

@end


@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self.sessionManager startPeerInvitationsWithServiceType:@"pinchcanvas" errorHandler:^(NSError *error) {
        NSLog(@"invitations failed with error %@", error);
    }];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self.sessionManager stopPeerInviations];
}

- (BOOL)prefersStatusBarHidden {
    return YES;
}

- (void)updateLabel {
    NSUInteger numberOfConnectedPeers = self.sessionManager.session.connectedPeers.count;
    NSUInteger numberOfConnectedScreens = [SCLScreen mainScreen].connectedScreens.count;
    
    self.textLabel.text = [NSString stringWithFormat:@"%d of %d screens connected", (int)numberOfConnectedScreens, (int)numberOfConnectedPeers];
}

- (void)updateImageFrame {
    CGFloat angle = 0.0;
    CGRect  frame = self.view.bounds;
    
    SCLScreen *localScreen = [SCLScreen mainScreen];
    if (localScreen.layout != nil) {
        // align image rotation to the first screen in the layout
        SCLScreen *originScreen = localScreen.layout.screens.firstObject;
        angle = [originScreen convertAngle:0 toCoordinateSpace:self.view];
        
        // extend image frame to the entire bounds of the layout
        frame = [localScreen.layout boundsInScreen:localScreen];
        frame = [localScreen convertRect:frame toCoordinateSpace:self.view];
    }
    
    [UIView animateWithDuration:0.25 animations:^{
        self.imageView.transform = CGAffineTransformMakeRotation(angle);
        self.imageView.frame = frame;
    }];
}

#pragma mark - MCSessionDelegate

// remote peer changed state
- (void)session:(MCSession *)session peer:(MCPeerID *)peerID didChangeState:(MCSessionState)state {
    switch (state) {
        case MCSessionStateNotConnected:
            NSLog(@"peer not connected: %@", peerID);
            break;
        case MCSessionStateConnecting:
            NSLog(@"peer connecting: %@", peerID);
            break;
        case MCSessionStateConnected:
            NSLog(@"peer connected: %@", peerID);
            break;
    }
    
    [self updateLabel];
}

#pragma mark - SCLLayoutObserving

// screen layout changed
- (void)layoutDidChangeForScreens:(NSArray *)updatedScreens {
    NSLog(@"layout changed for screens: %@", updatedScreens);
    
    if ([updatedScreens containsObject:[SCLScreen mainScreen]]) {
        [self updateLabel];
        [self updateImageFrame];
    }
}

@end

File: Examples/PinchCanvasSwift/PinchCanvasSwift/ViewController.swift
//
//  ViewController.swift
//  PinchCanvasSwift
//
//  Created by Jun on 12/10/14.
//  Copyright (c) 2014 Jun Tanaka. All rights reserved.
//

import UIKit
import ScreenLayout

class ViewController: SCLPinchViewController {
    
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var textLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        self.sessionManager.startPeerInvitations(withServiceType: "pinchcanvas", errorHandler: { (error) -> Void in
            print("invitations failed with error: \(error)")
        })
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        self.sessionManager.stopPeerInviations();
    }
    
    override var prefersStatusBarHidden : Bool {
        return true;
    }
    
    func updateLabels() {
        let numberOfConnectedPeers = self.sessionManager.session.connectedPeers.count
        let numberOfConnectedScreens = SCLScreen.main().connectedScreens.count;
        
        self.textLabel.text = "\(numberOfConnectedScreens) of \(numberOfConnectedPeers) screens connected";
    }
    
    func updateImageFrame() {
        var angle: CGFloat = 0.0
        var frame: CGRect = self.view.bounds
        
        let localScreen = SCLScreen.main()
        if let layout = localScreen.layout {
            // align image rotation to the first screen in the layout
            let originScreen: SCLScreen = layout.screens.first!
            angle = originScreen.convertAngle(0.0, to: self.view)

            // extend image frame to the entire bounds of the layout
            frame = layout.bounds(in: localScreen)
            frame = localScreen.convert(frame, to: self.view)
        }
        
        UIView.animate(withDuration: 0.25, animations: { () -> Void in
            self.imageView.transform = CGAffineTransform(rotationAngle: angle)
            self.imageView.frame = frame
        })
    }
    
    // remote peer changed state
    override func session(_ session: MCSession, peer peerID: MCPeerID, didChange state: MCSessionState) {
        switch state {
        case .notConnected:
            print("peer not connected: \(peerID)")
        case .connecting:
            print("peer connecting: \(peerID)")
        case .connected:
            print("peer connected: \(peerID)")
        }
        
        self.updateLabels()
    }
    
    // screen layout changed
    override func layoutDidChange(for affectedScreens: [SCLScreen]) {
        self.updateLabels()
        self.updateImageFrame()
    }
    
}

